### PR TITLE
add before run task for different artifacts

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/MavenRunTaskUtil.java [hunk 1]
```diff
@@ -68,10 +68,14 @@ public static void addMavenPackageBeforeRunTask(RunConfiguration runConfiguratio
     public static List<Artifact> collectProjectArtifact(@NotNull Project project) {
         List<Artifact> artifacts = new ArrayList<>();
         ArtifactType warArtifactType = ArtifactType.findById(MavenConstants.TYPE_WAR);
+        ArtifactType earArtifactType = ArtifactType.findById("ear");
         ArtifactType jarArtifactType = ArtifactType.findById(MavenConstants.TYPE_JAR);
         if (warArtifactType != null) {
             artifacts.addAll(ArtifactManager.getInstance(project).getArtifactsByType(warArtifactType));
         }
+        if (earArtifactType != null) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/MavenRunTaskUtil.java:76]: create a new method `collectArtifactByType` for the common logic?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                ExternalSystemBeforeRunTask task = provider.createTask(runConfiguration);
+                task.getTaskExecutionSettings()
+                    .setExternalSystemIdString(GradleConstants.SYSTEM_ID.toString());
+                task.getTaskExecutionSettings().setExternalProjectPath(gradleProject.getPath());
+                task.getTaskExecutionSettings().setTaskNames(Arrays.asList(GRADLE_TASK_ASSEMBLE));
+                return task;
+            }, runConfiguration, add);
+    }
+
+    public static <T extends BeforeRunTask> void updateBeforeRunOption(
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java:150]: change to `addOrRemoveBuildGradleProjectBeforeRunOption` to keep consistent with previous methods?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            List<T> tasks = ContainerUtil.findAll(editor.getStepsBeforeLaunch(), runTaskClass)
+                                         .stream()
+                                         .filter(filter)
+                                         .collect(Collectors.toList());
+            if (add && tasks.isEmpty()) {
+                T task = producer.produce();
+                task.setEnabled(true);
+                tasks.add(task);
+
+                RunManagerEx manager = RunManagerEx.getInstanceEx(runConfiguration.getProject());
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java:170]: Please comment why we need to add the `task2`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                dataContext);
+        if (editor != null) {
+            List<T> tasks = ContainerUtil.findAll(editor.getStepsBeforeLaunch(), runTaskClass)
+                                         .stream()
+                                         .filter(filter)
+                                         .collect(Collectors.toList());
+            if (add && tasks.isEmpty()) {
+                T task = producer.produce();
+                task.setEnabled(true);
+                tasks.add(task);
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java:168]: do we need to add `task` to `tasks` here? seems no more reference

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/components/AzureArtifactManager.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                return null;
+        }
+    }
+
+    private String getGradleProjectId(ExternalProjectPojo gradleProjectPojo) {
+        ExternalProject externalProject = getRelatedExternalProject(gradleProjectPojo);
+        return Objects.nonNull(externalProject) ? externalProject.getQName() : null;
+    }
+
+    private ExternalProject getRelatedExternalProject(ExternalProjectPojo gradleProjectPojo) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/components/AzureArtifactManager.java:115]: `getRelatedExternalProject` -> `getGradleProject`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java (hunk 4 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java:168]: reasonable.

Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java (hunk 3 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/BeforeRunTaskUtils.java:165]: Done

**Review by Flanker32 — APPROVED** [review]

**Comment by andxu** [comment]
Replaced with https://github.com/microsoft/azure-tools-for-java/pull/4636

**Pull request closed without merging**